src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,33 +127,177 @@ def __init__(
+
+    def _get_billing_producer(self) -> KafkaProducer:
+        servers = _get_bootstrap_servers(topic=self._billing_topic.name, force_cluster=None)
+        return KafkaProducer(
+            build_kafka_configuration(default_config={}, bootstrap_servers=servers)
+        )
 
     def poll(self) -> None:
         if self._should_commit():
             self._bulk_commit()
 
+    def _should_commit(self) -> bool:
+        self._mark_commit_ready()
+
+        if not self._ready_to_commit:
+            return False
+        if self._messages_ready_since_last_commit >= self._max_batch_size:
+            return True
+        if self._last_commit_time + self._max_batch_time <= datetime.now():
+            return True
+        return False
+
+    def _mark_commit_ready(self) -> None:
+        """Removes completed futures at the beginning of
+        ongoing_billing_outcomes and adds them to ready_to_commit.
+
+        If a future has errored, log an error but add to ready_to_commit
+        anyways.
+        """
+        while self._ongoing_billing_outcomes:
+            produce_billing, metrics_msg = self._ongoing_billing_outcomes[0]
+
+            if produce_billing:
+                if not produce_billing.done():
+                    break
+                # XXX(iker): in tests .exception() may return a mock, which
+                # resolves to a truthy value. Explicitly checking if it's an
+                # exception removes this log from tests. We may want to improve
+                # mocking in the tests and only check for a None value here.
+                if isinstance(produce_billing.exception(), Exception):
+                    logger.error(
+                        "Async future failed in billing metrics consumer.",
+                        exc_info=produce_billing.exception(),
+                        extra={"offset": metrics_msg.offset},
+                    )
+
+            self._ongoing_billing_outcomes.popleft()
+            self._ready_to_commit[metrics_msg.partition] = Position(
+                metrics_msg.next_offset, datetime.now()
+            )
+            self._messages_ready_since_last_commit += 1
+
+    def _bulk_commit(self) -> None:
+        """Commits and clears the ready to commit queue."""
+        if not self._ready_to_commit:
+            return
+        metrics.gauge(
+            "billing_metrics_consumer.bulk_commit.queue_size",
+            value=self._messages_ready_since_last_commit,
+        )
+        self._commit(self._ready_to_commit)
+        self._clear_ready_queue()
+
+    def _clear_ready_queue(self) -> int:
+        """Clears the ready to commit queue and returns the amount of items dropped."""
+        amount_dropped = self._messages_ready_since_last_commit
+        self._ready_to_commit = {}
+        self._messages_ready_since_last_commit = 0
+        self._last_commit_time = datetime.now()
+        return amount_dropped
+
+    def join(self, timeout: Optional[float] = None) -> None:
```

Comment:
Since you create the producer in the constructor of this strategy you will have to close it in the join method. 
Though this will not be needed if you follow the advice Lyn gave you and relied on the `ProduceAndCommit` strategy that does all this logic for you.